commands2/sysid/sysidroutine.py

```diff
@@ -0,0 +1,177 @@
+        recordState: Optional[Callable[[State], None]] = None
+
+    @dataclass
+    class Mechanism:
+        """Create a new mechanism specification for a SysId routine.
+
+        :param drive:     Sends the SysId-specified drive signal to the mechanism motors during test
+                          routines.
+        :param log:       Returns measured data of the mechanism motors during test routines. To return
+                          data, call `motor(string motorName)` on the supplied `SysIdRoutineLog` instance, and then
+                          call one or more of the chainable logging handles (e.g. `voltage`) on the returned
+                          `MotorLog`. Multiple motors can be logged in a single callback by calling `motor`
+                          multiple times.
+        :param subsystem: The subsystem containing the motor(s) that is (or are) being characterized.
+                          Will be declared as a requirement for the returned test commands.
+        :param name:      The name of the mechanism being tested. Will be appended to the log entry title
+                          for the routine's test state, e.g. "sysid-test-state-mechanism". Defaults to the name of
+                          the subsystem if left null.
+        """
+
+        drive: Callable[[volts], None]
+        log: Callable[[SysIdRoutineLog], None]
+        subsystem: Subsystem
+        name: Optional[str] = None
+
+        def __post_init__(self):
+            if self.name == None:
+                self.name = self.subsystem.getName()
+
+    class Direction(Enum):
+        """Motor direction for a SysId test."""
+
+        kForward = 1
+        kReverse = -1
+
+    def __init__(self, config: Config, mechanism: Mechanism):
+        """Create a new SysId characterization routine.
+
+        :param config:    Hardware-independent parameters for the SysId routine.
+        :param mechanism: Hardware interface for the SysId routine.
+        """
+        super().__init__(mechanism.subsystem.getName())
+        self.config = config
+        self.mechanism = mechanism
+        self.outputVolts = 0.0
+        self.logState = config.recordState or self.recordState
+
+    def quasistatic(self, direction: Direction) -> Command:
+        """Returns a command to run a quasistatic test in the specified direction.
+
+        The command will call the `drive` and `log` callbacks supplied at routine construction once
+        per iteration. Upon command end or interruption, the `drive` callback is called with a value of
+        0 volts.
+
+        :param direction: The direction in which to run the test.
+
+        :returns: A command to run the test.
+        """
+
+        timer = Timer()
+        output_sign = 1.0 if direction == self.Direction.kForward else -1.0
+        state = {
+            self.Direction.kForward: State.kQuasistaticForward,
+            self.Direction.kReverse: State.kQuasistaticReverse,
+        }[direction]
```

Comment:
These changed upstream from using a dict to an if statement. https://github.com/wpilibsuite/allwpilib/pull/6209